### ISSUE TITLE
When expressions & workspace specific commands

### ISSUE THREAD

**Issue by elazarcoh** [usernamehw/vscode-commands#20]
Hi @usernamehw,
I had the same idea for extension some time ago.
I've started toying with it. In my initial implementation (which I never published), I had some feature that I want to contribute here, if you would accept it:
1. Support `when` expression for the command-palette. This gives more flexibility of when to show command in the command palette, for example.
2. Support workspace specific commands. Those commands would be available only in the workspace they where written in.
3. Implement new commands in a js file, and dynamically run it. This can be very convenient for creating a new command without having to create a new extension. 
    (This feature I have only in a POC state, so I won't expect to make a PR for it just yet)

Let me know what you think. would you accept PR for any of those?
We can discuss detail further, if you'd like.

**Comment by usernamehw** [comment]
I think 3rd one is an overkill. It would likely be easier to create an extension than to write a js file.

How would 2nd work? I think settings that are objects or arrays are merged (global + workspace), unless they were fetched by `inspect()` instead of `get()`. It might make more sense to have a separate setting just for workspace config.

I would accept PR for 1st & 2nd if they are relatively small/simple.

**Comment by elazarcoh** [comment]
> How would 2nd work? I think settings that are objects or arrays are merged (global + workspace), unless they were fetched by inspect() instead of get(). It might make more sense to have a separate setting just for workspace-specific commands.

I did have 2 different settings to accomplish this (otherwise I can't get both of the to work in the same time):
1. Global, with `machine` scope. 
2. Workspace, with `window` scope.

To support it, I added a unique id for each workspace, and checked it with a `when` expression.

OK, I'll start with a PR to support the `when` expression, and I'll add the workspace on top of it.

Thanks.
